REGRESSION(281440@main): Web content can become inaccessible after dynamic webpage changes

twilco opened the pull request:
#### 71ce3f1d011ad76b976f01882cf2a81129815ff9
<pre>
REGRESSION(281440@main): Web content can become inaccessible after dynamic webpage changes
<a href="https://rdar.apple.com/133477773">rdar://133477773</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=277817">https://bugs.webkit.org/show_bug.cgi?id=277817</a>

Reviewed by Andres Gonzalez and Chris Fleizach.

In <a href="http://commits.webkit.org/281440@main">http://commits.webkit.org/281440@main</a>, we fixed a bug where `AXIsolatedTree::m_protectedFromDeletionIDs` was
cleared too early, which in turn caused objects to be incorrectly removed from `AXIsolatedTree::m_nodeMap`, eventually
breaking the accessibility tree. We fixed this by moving:

`m_pendingProtectedFromDeletionIDs.formUnion(std::exchange(m_protectedFromDeletionIDs, { }))`

from `queueRemovalsLocked` to `queueAppendsAndRemovals` (which should happen once per tree-update cycle, making it the
right time to clear this list).

This caused a new bug, however, in this sequence:

  1. The main-thread starts a tree update via `AXIsolatedTree::updateChildren`.

  2. In doing so, we call `queueRemovalsLocked`, i.e. because an object lost a child (but that child isn&apos;t deleted, just
     has a different parent, so we mark it as protected on the main-thread in `m_protectedFromDeletionIDs`).

  3. We call `queueRemovalsLocked` for that child and any others no longer children of the original object. After
     281440@main, we no longer sync protected IDs in this function.

  4. Before the full main-thread tree-update (`AXIsolatedTree::updateChildren`) is finished, resulting in a call to
     `queueAppendsAndRemovals`, the secondary thread processes a request from an AT, causing `AXIsolatedTree::applyPendingChanges`
     to run on the secondary thread (as expected to ensure we service the request with the most up-to-date information).

  5. We delete the re-parented child, and any of its descendants, from the accessibility thread data structures because
     we failed to protect it.

  6. `queueAppendsAndRemovals` eventually finishes on the main-thread, and we sync the protected objects to the secondary
     thread, but at that point it&apos;s too late.

  7. The accessibility tree is now broken, with random objects missing entirely.

This patch fixes this by changing `queueRemovalsLocked` (called by `queueAppendsAndRemovals`) to sync protected object
IDs if necessary, and changes `queueAppendsAndRemovals` to clear the list of protected objects, fixing both the bug
addressed by <a href="http://commits.webkit.org/281440@main">http://commits.webkit.org/281440@main</a>, and the bug it caused.

Because 281440@main caused a bug, it was reverted in <a href="https://github.com/WebKit/WebKit/pull/31917.">https://github.com/WebKit/WebKit/pull/31917.</a>
This PR brings back the test 281440@main added (animated-dropdown.html), and adds a new test (child-update-during-ax-request.html)
that would&apos;ve caught the bug 281440@main introduced.

* LayoutTests/accessibility/animated-dropdown-expected.txt: Added.
* LayoutTests/accessibility/animated-dropdown.html: Added.
* LayoutTests/accessibility/mac/child-update-during-ax-request-expected.txt: Added.
* LayoutTests/accessibility/mac/child-update-during-ax-request.html: Added.
* LayoutTests/accessibility/resources/jquery-3.6.1.js: Added.
* LayoutTests/platform/glib/TestExpectations: Skip accessibility/animated-dropdown.html.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::queueRemovalsLocked):
(WebCore::AXIsolatedTree::queueAppendsAndRemovals):
(WebCore::AXIsolatedTree::collectNodeChangesForSubtree):
(WebCore::AXIsolatedTree::updateChildren):
(WebCore::AXIsolatedTree::protectFromDeletion):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/282266@main">https://commits.webkit.org/282266@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/634b3826a80280bba6a2261dedf2c739a1ffa618

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62596 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41951 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15191 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66580 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13148 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64716 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49638 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13484 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50434 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9067 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65665 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38987 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54212 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31176 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35709 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11547 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12076 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57295 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11863 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68311 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6542 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11551 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57810 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6572 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54259 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58008 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13902 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5459 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37751 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38836 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39933 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38580 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on LayoutTests/accessibility/mac/child-update-during-ax-request.html:
```diff
@@ -0,0 +1,141 @@
+<script>
+var output = "This test ensures that the accessiblity tree doesn't become broken when accessible children change in the middle of servicing an AT request.\n\n";
+
+// This test must *never* be flaky. If there is a bug in the implementation, it will fail only sometimes. If this test
```
Not for right now.

Rather than a comment, it might be helpful to have a directory of manual tests, each one could have instructions for what to do to reproduce the bug. That could be helpful for some of the active descendant stuff too.

twilco replied:
Good idea! Filed https://bugs.webkit.org/show_bug.cgi?id=277821 and added a FIXME to this test that we can address in a follow-up.

minorninth commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -369,6 +369,8 @@ void AXIsolatedTree::queueRemovalsLocked(Vector<AXID>&& subtreeRemovals)
 
     m_pendingSubtreeRemovals.appendVector(WTFMove(subtreeRemovals));
+    if (m_protectedFromDeletionIDsIsDirty)
+        m_pendingProtectedFromDeletionIDs.formUnion(m_protectedFromDeletionIDs);
```
This line is just slightly confusing, do we really need most of the ids to end up in both sets?

I notice elsewhere you used std::exchange to delete from m_protectedFromDeletionIDs and union with m_pendingProtectedFromDeletionIDs, and I just want to make sure that wouldn't make sense here.

m_pendingProtectedFromDeletionIDs.formUnion(std::exchange(m_protectedFromDeletionIDs, { }));

twilco replied:
> This line is just slightly confusing, do we really need most of the ids to end up in both sets?

Right now, yes — it follows the pattern of other `AXIsolatedTree::m_pending*` data structures we sync for isolated tree updates, and these specific two are required to make sure we don't delete objects from the accessibility thread data structures, nor the main-thread data structures (`AXIsolatedTree::m_nodeMap`) that help us make correct and efficient updates to the tree. But overall this whole system is over-complicated — I have ideas I'm pursuing this upcoming year to simplify things, so hopefully some of this can be removed outright.

> I notice elsewhere you used std::exchange to delete from m_protectedFromDeletionIDs and union with m_pendingProtectedFromDeletionIDs, and I just want to make sure that wouldn't make sense here.

Indeed, this is actually what we did prior to http://commits.webkit.org/281440@main. But it caused `m_protectedFromDeletionIDs` to be cleared too early, before a full tree-update cycle, causing the `m_nodeMap` to be updated incorrectly.

minorninth commented on LayoutTests/accessibility/mac/child-update-during-ax-request.html:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    // Every traversal must include these elements. If not, it's very likely the accessibility tree is broken.
+    if (tableCount != 1 || cellCount != 4 || rowCount != 2 || buttonCount != 2 || !foundThirdCellText || !foundFourthCellText) {
```
Could it just diff searchOutput relative to the initial rather than counting a bunch of things? Or does that make it harder to debug?

hoffmanjoshua replied:
This is a personal preference, but I find the failure logs that include counts + bools (like this) easier to diagnose, and a quick way to see why a test might be failing. Diffs don't make that immediately clear.

twilco replied:
> Could it just diff searchOutput relative to the initial rather than counting a bunch of things?

Normally I would do this, but because this test constantly adding random strings to the webpage, and hiding and un-hiding a div with text, the `searchOutput` diff could be different in non-important ways. I also figured this would be faster than doing a bunch of JavaScript string `includes` checks, as the output string is large, and this test is already a bit slow. But certainly not set in stone if you have another idea!

minorninth reviewed (commented):
(Not a WebKit reviewer.)

lgtm

One logic question and a couple of comments, but no concerns. The test has some clever tricks, nice job.

AndresGonzalezApple commented on LayoutTests/accessibility/animated-dropdown.html:
```diff
@@ -0,0 +1,91 @@
+    output += dumpAXSearchTraversal(accessibilityController.rootElement.childAtIndex(0));
+    var secondTraversal;
+    setTimeout(async function() {
+        await sleep(1000)
```
Why do we need to sleep here?

twilco replied:
Oops, I think this is left over from when I was tweaking this test. Even if a sleep is necessary, surely it can be shorter than one second while still triggering the bug-under-test. Let me go back and look into this.

twilco replied:
The `sleep` can be removed outright without affecting whether the bug is triggered, so I've done so in the latest commit.

AndresGonzalezApple commented on LayoutTests/accessibility/animated-dropdown.html:
```diff
@@ -0,0 +1,91 @@
+        o.O.j = t => 0 === a[t];
+    })(), o.O(void 0, [], (() => o(641))), o.O();
+    var s = o.O(s);
+})();
```
What is this?

twilco replied:
Minified JS taken from a webpage exhibiting this bug. Originally the corpus was significantly larger, but I spent a ton of time reducing it to this — I couldn't get it any smaller without breaking the behavior under-test.

AndresGonzalezApple replied:
Greatly appreciate that you took the big pain of reducing the test case. I'm still confused about this block of code with lots of 1 letter identifiers and style that don't follow WebKit standards. Can we comment this, or even write it in a more readable way?

twilco replied:
OK, all of that minified JS is removed in the latest commit.

AndresGonzalezApple commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
```diff
@@ -491,6 +492,9 @@ class AXIsolatedTree : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<AX
+    bool m_protectedFromDeletionIDsIsDirty { false };
     // Only accessed on the main thread.
     // Objects whose parent has changed, and said change needs to be synced to the secondary thread.
     HashSet<AXID> m_needsParentUpdate;
```
This could be greatly simplified if we move away from calling applyPendingChanges for every AX thread request, but instead we update the iso tree on the main thread holding the lock. The problem of serving an AX thread request during an update should disappear altogether.

twilco replied:
Seems like a good follow-up change.

twilco committed:
REGRESSION(281440@main): Web content can become inaccessible after dynamic webpage changes
rdar://133477773
https://bugs.webkit.org/show_bug.cgi?id=277817

Reviewed by Andres Gonzalez and Chris Fleizach.

In http://commits.webkit.org/281440@main, we fixed a bug where `AXIsolatedTree::m_protectedFromDeletionIDs` was
cleared too early, which in turn caused objects to be incorrectly removed from `AXIsolatedTree::m_nodeMap`, eventually
breaking the accessibility tree. We fixed this by moving:

`m_pendingProtectedFromDeletionIDs.formUnion(std::exchange(m_protectedFromDeletionIDs, { }))`

from `queueRemovalsLocked` to `queueAppendsAndRemovals` (which should happen once per tree-update cycle, making it the
right time to clear this list).

This caused a new bug, however, in this sequence:

  1. The main-thread starts a tree update via `AXIsolatedTree::updateChildren`.

  2. In doing so, we call `queueRemovalsLocked`, i.e. because an object lost a child (but that child isn't deleted, just
     has a different parent, so we mark it as protected on the main-thread in `m_protectedFromDeletionIDs`).

  3. We call `queueRemovalsLocked` for that child and any others no longer children of the original object. After
     281440@main, we no longer sync protected IDs in this function.

  4. Before the full main-thread tree-update (`AXIsolatedTree::updateChildren`) is finished, resulting in a call to
     `queueAppendsAndRemovals`, the secondary thread processes a request from an AT, causing `AXIsolatedTree::applyPendingChanges`
     to run on the secondary thread (as expected to ensure we service the request with the most up-to-date information).

  5. We delete the re-parented child, and any of its descendants, from the accessibility thread data structures because
     we failed to protect it.

  6. `queueAppendsAndRemovals` eventually finishes on the main-thread, and we sync the protected objects to the secondary
     thread, but at that point it's too late.

  7. The accessibility tree is now broken, with random objects missing entirely.

This patch fixes this by changing `queueRemovalsLocked` (called by `queueAppendsAndRemovals`) to sync protected object
IDs if necessary, and changes `queueAppendsAndRemovals` to clear the list of protected objects, fixing both the bug
addressed by http://commits.webkit.org/281440@main, and the bug it caused.

Because 281440@main caused a bug, it was reverted in https://github.com/WebKit/WebKit/pull/31917.
This PR brings back the test 281440@main added (animated-dropdown.html), and adds a new test (child-update-during-ax-request.html)
that would've caught the bug 281440@main introduced.

* LayoutTests/accessibility/animated-dropdown-expected.txt: Added.
* LayoutTests/accessibility/animated-dropdown.html: Added.
* LayoutTests/accessibility/mac/child-update-during-ax-request-expected.txt: Added.
* LayoutTests/accessibility/mac/child-update-during-ax-request.html: Added.
* LayoutTests/accessibility/resources/jquery-3.6.1.js: Added.
* LayoutTests/platform/glib/TestExpectations: Skip accessibility/animated-dropdown.html.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::queueRemovalsLocked):
(WebCore::AXIsolatedTree::queueAppendsAndRemovals):
(WebCore::AXIsolatedTree::collectNodeChangesForSubtree):
(WebCore::AXIsolatedTree::updateChildren):
(WebCore::AXIsolatedTree::protectFromDeletion):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/282266@main
